syntax = "proto3";

message PutRequest {
    uint32 key = 1;
    string value = 2;
    uint32 consistency_level = 3; 
}

message GetRequest {
    uint32 key = 1;
    uint32 consistency_level = 2;
    string value = 3;
    double timestamp = 4;
}

message CordResponse { 
    uint32 id = 1;
    uint32 key = 2;
    bool status = 3;        
    string value = 4;
    double timestamp = 5;
}

message ReplicaResponse {
    uint32 id = 1;
    uint32 key = 2;
    string value = 3;
    bool status = 4;        
    double timestamp = 5;    // unix timestamp;. String okay? Or int32 or int64?
}

message ReplicaRequest {
    uint32 key = 1;
    string value = 2;
    double timestamp = 3;
    uint32 operation = 4;    // 0--> Read, 1-->Write
}

message ErrorMessage {
    string msg = 1;
}

message ReadRepair{
    uint32 key = 1;
    string value = 2;
    double timestamp = 3; 
}

message HintedHandoff{
    uint32 key = 1;
    string value = 2;
    double timestamp = 3; 
}

message KVMessage {
  oneof kv_message {
    PutRequest put_request = 1;
    GetRequest get_request = 2;
    CordResponse cord_response = 3;
    ReplicaRequest replica_request = 4;
    ReplicaResponse replica_response = 5;
    ErrorMessage error_message = 6;
  }
}